### PR TITLE
fix issue 16542 -  makeArray not usable with const initializer

### PR DESCRIPTION
**Pull request by somzzz**
Let me know if this is a suitable solution, thanks!

### CODE REVIEW

**Comment by BBasile** [comment]
Add a unittest please, because the issue looks invalid:

```d
unittest
{
    const int[] v;
    void test(T)(in T t)
    {
        auto a = theAllocator.makeArray2(100, t);
    }
    test(v);
}
```

passes without removing `auto ref`.

**Comment by somzzz** [comment]
The dmd version i tested with is DMD64 D Compiler ```v2.073.0-devel-538a895```.

The unittest passes, but compilation fails for this program:
````
import std.experimental.allocator;

void test(T)(in T initialValue)
{
      auto t = theAllocator.makeArray!T(100, initialValue); // won't compile
      auto t2 = theAllocator.makeArray!T(100, cast()initialValue); // this one compiles
}

int main()
{
    const int init = 3;
    const int[] v;
   // test(init);
    test(v);
    return 0;
}
````

**Comment by BBasile** [comment]
The problem is a wrong template parameter. With inference it works:

```d
void test(T)(in T initialValue)
{
      auto t = theAllocator.makeArray(100, initialValue); 
}
```

The issue looks well invalid.

**Comment by andralex** [comment]
@somzzz could you then convert your test program into a unittest that fails without the change and passes with it? thx!

**Comment by JackStouffer** [comment]
Ping @somzzz

**Comment by somzzz** [comment]
The ```unittest``` doesn't compile without the change. It's the example in the bug report.

File: std/experimental/allocator/package.d [hunk 1]
```diff
@@ -804,8 +804,7 @@ private enum hasPureDtor(T) = !hasElaborateDestructor!T ||
 private enum canSafelyDeallocPostRewind(T) = hasPurePostblit!T && hasPureDtor!T;
 
 /// Ditto
-T[] makeArray(T, Allocator)(auto ref Allocator alloc, size_t length,
-    auto ref T init)
+T[] makeArray(T, Allocator)(auto ref Allocator alloc, size_t length, T init)
```
Comment (wilzbach) [inline std/experimental/allocator/package.d]: Why don't you use `in` or `inout` here? Passing parameters by reference if possible is usually faster.

Comment (andralex) [reply std/experimental/allocator/package.d]: @wilzbach which parameter do you mean?

Comment (ZombineDev) [reply std/experimental/allocator/package.d]: @andralex my guess would be `T init` -> `in T init`

**Comment by MetaLang** [comment]
vibe.d tests are failing on Jenkins, but I can't tell if this PR broke it. 

```
Running ./tests 
Watchdog receiver 1
Watchdog receiver 2
send Hello World
Received string message: Hello, World!
Watchdog receiver 3
Watchdog receiver 4
Watchdog receiver 5
send int 1
Watchdog receiver 6
Received double: 1
Watchdog receiver 7
Watchdog receiver 8
Watchdog receiver 9
send double 1.2
Watchdog receiver 10
Watchdog receiver 11
Watchdog receiver 12
Watchdog receiver 13
send int 2
Received int message: 2
Received double: 1.2
Watchdog receiver 14
Watchdog receiver 15
Watchdog receiver 16
Watchdog receiver 17
send 3xint 1 2 3
Watchdog receiver 18
Watchdog receiver 19
Watchdog receiver 20
Watchdog receiver 21
send string Bye bye
Received string message: Bye bye
Received iii: 1 2 3
Receive loop finished.
Receiver watchdog failure.
uncaught exception
dwarfeh(224) fatal error
Program exited with code -6
script returned exit code 2
```

File: std/experimental/allocator/package.d (hunk 1 above)
Comment (andralex) [reply std/experimental/allocator/package.d]: `in`/`inout` don't pass by reference, and the previous code with ref didn't work. I'll pull this as is, improvements are welcome

**Comment by somzzz** [comment]
Looks like the style checker segfaults? I'm not sure what this means.
```
Running DScanner
../dscanner-285ef162f024cbd305d587e9e0fcfb2292ea93ce/dsc --config .dscanner.ini --styleCheck etc std -I.
make: *** [dscanner] Segmentation fault (core dumped)
```

**Comment by andralex** [comment]
@wilzbach can you please give it a look? thx!

**Comment by andralex** [comment]
will force pull for now

**Pull request merged**